### PR TITLE
Remove any types and implement proper type assertions

### PR DESCRIPTION
**Pull request by AMoldskred**
In this PR I have started to remove any-types by creating proper type-asssertions.
I have also moved some conditionals to make it both easier to read and for typescript to easier infer the correct type.

It is not a complete fix, but it should be a good start.
Before merging with master I would try to pull locally and test that it doesn't break anything.
I couldn't figure out how to use the correct ts-lint and how to run it. But I am fairly certain that it should be all good

#14

### CODE REVIEW

File: packages/custom-elements-json/schema.ts [hunk 1]
```diff
@@ -130,6 +130,7 @@ export type Declaration =
   | ClassDeclaration
   | FunctionDeclaration
   | VariableDeclaration
+  | MixinDeclaration
```
Comment (thepassle) [inline packages/custom-elements-json/schema.ts:133]: We should PR this here https://github.com/webcomponents/custom-elements-json/pull/9 as well

File: packages/custom-elements-json-core/src/utils/index.ts [hunk 2]
```diff
@@ -172,12 +185,20 @@ export function hasAggregatingImport(node: any): boolean {
...
+  kind: 'default' | 'named' | 'aggregate' | 'javascript-module';
   importPath: string;
-  isBareModuleSpecifier: boolean
-}
\ No newline at end of file
+  isBareModuleSpecifier: boolean;
+}
+
+/** Type asserters */
+export function isImport<T>(C: T | Import): C is Import {
```
Comment (thepassle) [inline packages/custom-elements-json-core/src/utils/index.ts:219]: My typescript-foo obviously isnt very strong, could you provide some details on this syntax and how it works exactly?

`isImport<T>(C: T | Import): C is Import {`

**Comment by thepassle** [comment]
This is really great! Very valuable help here, I appreciate it a lot 🙂 

It does seem like theres some breakage regarding the types however:
image

Could you take a look? You can confirm the types are working by, from the package root:

```bash
yarn install
cd packages/custom-elements-json-core
npm run tsc:watch
```

**Comment by AMoldskred** [comment]
Will do!

File: packages/custom-elements-json-core/src/utils/index.ts (hunk 2 above)
Comment (AMoldskred) [reply packages/custom-elements-json-core/src/utils/index.ts:219]: Basically what it does is to take in an argument which can be of type Import or some thing else (T). And then we declare that based on the boolean we return in the function we can assert wether or not C is of type Import. 

So whenever we use this in a if-statement typescript will understand that the variable in question is in fact of type Import.

Comment (thepassle) [reply packages/custom-elements-json-core/src/utils/index.ts:219]: TIL, thanks!

**Comment by AMoldskred** [comment]
@thepassle I'm not going to lie. There needs some serious work done to make typing to work properly in this project.
I have done what I can for now to fix the typing, however there is one place where I had to force any type in `src/ast/handleEvents` (There was no logical way that I could see that would infer the correct type).

**Comment by AMoldskred** [comment]
@thepassle It would be really great if you could tag this PR with `hacktoberfest-accepted` so that I could get this as a PR towards my 4 in hacktoberfest😉

**Comment by thepassle** [comment]
> @thepassle It would be really great if you could tag this PR with `hacktoberfest-accepted` so that I could get this as a PR towards my 4 in hacktoberfest😉

Sorry for the late reply, I've had some IRL stuff to attend to this past week, and I've also been considering the approach of this project, and potentially just moving it over to JS with JSDoc, but I'm still unsure on what to do.

Does adding a `hacktoberfest-accepted` immediately count towards your hacktober fest progress? Even if this doesn't get merged? If so, I'll happily add it so you can complete your hacktoberfest goal.

**Comment by AMoldskred** [comment]
Thank you @thepassle ! The PR started counting as soon as you added the label. I wish you best of luck with your project!

**Pull request closed without merging**